seed/csharp-sdk/pagination/src/SeedPagination/Core/Pager.cs

```diff
@@ -0,0 +1,211 @@
+    )
+    {
+        await foreach (var page in AsPagesAsync(cancellationToken).ConfigureAwait(false))
+        {
+            foreach (var value in page.Items)
+            {
+                yield return value;
+            }
+        }
+    }
+}
+
+internal sealed class OffsetPager<TRequest, TRequestOptions, TResponse, TOffset, TStep, TItem>
+    : Pager<TItem>
+{
+    private TRequest _request;
+    private readonly TRequestOptions? _options;
+    private readonly GetNextPage _getNextPage;
+    private readonly GetOffset _getOffset;
+    private readonly SetOffset _setOffset;
+    private readonly GetStep? _getStep;
+    private readonly GetItems _getItems;
+    private readonly HasNextPage? _hasNextPage;
+
+    internal delegate Task<TResponse> GetNextPage(
+        TRequest request,
+        TRequestOptions? options,
+        CancellationToken cancellationToken
+    );
+
+    internal delegate TOffset GetOffset(TRequest request);
+
+    internal delegate void SetOffset(TRequest request, TOffset offset);
+
+    internal delegate TStep GetStep(TRequest request);
+
+    internal delegate IReadOnlyList<TItem>? GetItems(TResponse response);
+
+    internal delegate bool? HasNextPage(TResponse response);
+
+    internal OffsetPager(
+        TRequest request,
+        TRequestOptions? options,
+        GetNextPage getNextPage,
+        GetOffset getOffset,
+        SetOffset setOffset,
+        GetStep? getStep,
+        GetItems getItems,
+        HasNextPage? hasNextPage
+    )
+    {
+        _request = request;
+        _options = options;
+        _getNextPage = getNextPage;
+        _getOffset = getOffset;
+        _setOffset = setOffset;
+        _getStep = getStep;
+        _getItems = getItems;
+        _hasNextPage = hasNextPage;
+    }
+
+    public override async IAsyncEnumerable<Page<TItem>> AsPagesAsync(
+        [EnumeratorCancellation] CancellationToken cancellationToken = default
+    )
+    {
+        var hasStep = false;
+        if (_getStep is not null)
+        {
+            hasStep = _getStep(_request) is not null;
+        }
+        var offset = _getOffset(_request);
+        var longOffset = Convert.ToInt64(offset);
+        bool hasNextPage;
+        do
+        {
+            var response = await _getNextPage(_request, _options, cancellationToken)
+                .ConfigureAwait(false);
+            var items = _getItems(response);
+            var itemCount = items?.Count ?? 0;
+            hasNextPage = _hasNextPage?.Invoke(response) ?? itemCount > 0;
+            if (items is not null)
+            {
+                yield return new Page<TItem>(items);
+            }
+
+            // If there is a step, we need to increment the offset by the number of items
+            if (hasStep)
+            {
+                longOffset += items?.Count ?? 1;
+            }
+            else
+            {
+                longOffset++;
+            }
+
+            // ensure there's a request object to set the offset on
```

Comment:
nit: I'd remove most of the comments in this file, especially given that it will be generated as a core utility.